File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int r = matrix.size(),
            c = matrix[0].size();
        
        if (target < matrix[0][0] || target > matrix[r - 1][c - 1]) return false;
        if (target == matrix[0][0] || target == matrix[r - 1][c - 1]) return true;
        int s = 0,
            e = r * c - 1;
        
        while (s <= e) {
            int m = s + (e - s)/2,
                mid = matrix[m / c][m % c];
            if (mid < target) {
                s = m + 1;
            } else if (mid > target) {
                e = m - 1;
            } else {
                return true;
            }
        }
        
        return false;
    }
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.rs
use std::cmp::Ordering::{Less,Greater,Equal};
impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        let m = matrix.len();
        let n = matrix[0].len();
        
        if target < matrix[0][0] || matrix[m - 1][n - 1] < target {
            // target is overmatched by matrix
            return false;
        }
        
        let (mut i, mut j) = (0, m * n - 1);
        
        while i <= j {
            let x = (i + (j - i)/2) as usize;
            let v = matrix[x/n][x%n];
            if target == v {
                return true;
            } else if target > v {
                i = x + 1;
            } else {
                j = x - 1;
            }
        }
        
        false
    }
}